retworkx-core/src/distancemap.rs

```diff
@@ -0,0 +1,88 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+// This module was originally copied and forked from the upstream petgraph
+// repository, specifically:
+// https://github.com/petgraph/petgraph/blob/0.5.1/src/dijkstra.rs
+// this was necessary to modify the error handling to allow python callables
+// to be use for the input functions for edge_cost and return any exceptions
+// raised in Python instead of panicking
+
+//! This module contains the [`DistanceMap`] trait which is used in
+//! [`shortest_path`].
+//!
+//! The trait allows the shortest path functions to support multiple
+//! return types.
+
+use std::hash::Hash;
+
+use petgraph::graph::IndexType;
+
+use crate::dictmap::*;
+use hashbrown::HashMap;
+
+pub trait DistanceMap<K, V> {
```

Comment:
Do we need a docstring for this trait? I'm not sure module docs are sufficient to document how this is used or what the expectations for it are.

Reply:
I took a look at the output for cargo doc, I think we should add a docstring here. When you click `DistanceMap` from the shortest path functions you get this page:

![Screenshot_2022-01-14_13-03-00](https://user-images.githubusercontent.com/2447371/149563526-28966b27-53d1-4125-9ac9-18a29177c906.png)

The module page is visible if you explicitly go to it, but without manually searching for it you probably won't see it. So adding a docstring explaining what this trait is used for and what the methods in the trait do is probably good (just in case someone wants to implement it for their own type).